qenspy/qens.py

```diff
@@ -194,10 +273,151 @@ def predict(self, q, w):
             quantile level k = 1, ..., K
         """
         # adjust w to handle missing values in q
-        w = super().broadcast_w_and_renormalize(q, w)
+        w = super().broadcast_w_and_renormalize(q, w['w'])
 
         # calculate weighted mean along the M axis for each combination of N, K
-        ensemble_q = tf.reduce_sum(tf.math.multiply_no_nan(q, w), axis = 2)
+        q_nans_replaced = q
+        q_nans_replaced[np.isnan(q_nans_replaced)] = 0.0
+        ensemble_q = tf.reduce_sum(tf.math.multiply_no_nan(q_nans_replaced, w), axis = 2)
 
         # return
         return ensemble_q
+
+
+class MedianQEns(QEns):
+    def __init__(self, bw_method):
+        self.bw_method = bw_method
+
+    def calc_bandwidth(self, q, w = None):
+        """
+        Calculate bandwidth
+
+        Parameters
+        ----------
+        q: 3D tensor with shape (N, K, M)
+            Predictive quantiles from component models where N is the number of
+            location/forecast date/horizon triples, M is number of models,
+            and K is number of quantile levels
+        w: 2D tensor with shape (K, M)
+            Model weight. Not required if self.bw_method is “unweighted”
+
+        Returns
+        -------
+        bw: 2D tensor with shape (N, K)
+            bandwidths
+        """
+        if self.bw_method == "silverman_unweighted":
+            M = q.shape[2]
+            w_unweighted = tf.broadcast_to(tf.constant([1/M], dtype = q.dtype),[q.shape[1], q.shape[2]])
+            broadcast_w = super().broadcast_w_and_renormalize(q, w_unweighted)
+        
+        elif self.bw_method == "silverman_weighted":
+            if w is None:
+                raise ValueError("Please provide w.")
+            M = w.shape[1]
+            broadcast_w = super().broadcast_w_and_renormalize(q, w)
+
+        # calculate weighted mean along the M axis for each combination of N, K but keep extra dims
+        weighted_mean = tf.reduce_sum(tf.math.multiply_no_nan(q, broadcast_w), axis = 2, keepdims=True)
```

Comment:
To support estimation, we may need to use a similar trick here as was done in the mean ensemble predict method above to replace nan values in q with 0's.